src/agent/src/sandbox.rs

```diff
@@ -403,6 +407,107 @@ impl Sandbox {
             }
         });
     }
+
+    #[instrument]
+    pub fn add_shared_mounts(&self, c: &LinuxContainer, mounts: &Vec<SharedMount>) -> Result<()> {
+        let shared_mounts = self.shared_mounts.clone();
+        let mounts = mounts.clone();
+        let logger = self.logger.clone();
+        let mut mnt_ns = Namespace::new(&self.logger).get_mount();
+        mnt_ns.path = format!("/proc/{}/ns/mnt", c.init_process_pid);
+
+        let new_thread = std::thread::spawn(move || {
+            if let Err(err) = || -> Result<()> {
+                mnt_ns.switch_to()?;
+                block_on(async move {
+                    let mut shared_mounts = shared_mounts.lock().await;
+                    for m in mounts {
+                        if m.type_() == "src" {
+                            fs::create_dir_all(&m.path)?;
+                            let source = Path::new(&m.path);
+                            let destination = Path::new(&m.path);
+                            baremount(source, destination, "bind", MsFlags::MS_BIND, "", &logger)?;
+                            mount(Some(""), destination, Some(""), MsFlags::MS_SHARED, Some(""))?;
+
+                            let src = std::ffi::CString::new(m.path)?;
+                            let mount_fd = unsafe {
+                                syscall(
+                                    libc::SYS_open_tree,
+                                    libc::AT_FDCWD,
+                                    src.as_ptr(),
+                                    1, // OPEN_TREE_CLONE
+                                ) as i32
+                            };
+                            
+                            shared_mounts.insert(m.name, mount_fd);
+                        }
+                    }
+
+                    Ok(())
+                })
+            }() {
+                return Err(err);
+            }
+
+            Ok(())
+        });
+        
+        new_thread
+            .join()
+            .map_err(|e| anyhow!("Failed to join thread {:?}!", e))??;
+
+        Ok(())
+    }
+
+    #[instrument]
+    pub fn setup_shared_mounts(&self, c: &LinuxContainer, mounts: &Vec<SharedMount>) -> Result<()> {
+        let shared_mounts = self.shared_mounts.clone();
+        let mounts = mounts.clone();
+        let mut mnt_ns = Namespace::new(&self.logger).get_mount();
+        mnt_ns.path = format!("/proc/{}/ns/mnt", c.init_process_pid);
+    
+        let new_thread = std::thread::spawn(move || {
+            if let Err(err) = || -> Result<()> {
+                mnt_ns.switch_to()?;
+                block_on(async move {
+                    for m in mounts {
+                        if m.type_() == "dst" {
+                            let shared_mounts = shared_mounts.lock().await;
```

Comment:
The thread may win the lock race before `add_shared_mounts` and find out `shared_mounts` to be empty. We need something more reliable to make sure `move_mount` happens **after** `open_tree`.

Actually, I think the current workflow is a bit confusing. Can we simplify it a little bit such that:
1. when creating the source mount point container, do nothing;
2. when creating the destination mount point container, if there are shared mount configured, do the following
  1. spawn a new thread, which
  2. waits for the source mount point in the source container to be ready, by polling the mount point directory to see if it becomes a mount point, with a short sleep between each poll of course
  3. when the source mount point is ready, create a bind mount from it to the destination mount point

Reply:
Thank you so much for your thoughtful review and valuable suggestions. I'll promptly make the necessary changes to the PR.